perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/BBMRIResources.java

```diff
@@ -0,0 +1,318 @@
+import cz.metacentrum.perun.registrar.model.Application;
+import cz.metacentrum.perun.registrar.model.ApplicationFormItemData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Registration module for BBMRI Collections, Networks, canSERV services, etc.
+ * Module:
+ * 1. reads input with IDs of the resources and checks, whether groups representing resources exist
+ *   - uses the content of attribute RESOURCE_ID_ATTR_NAME as the name of the attribute containing IDs of the
+ *     resources on groups
+ *   - parses the root group, under which the groups representing specified resources are located
+ *     - if RESOURCE_ORIGIN_ENABLED_ATTR_NAME is set to true, the root group will be parsed from input from the form
+ *       (where input is select named "resourceOrigin")
+ *     - else if RESOURCES_ROOT_GROUP_ATTR_NAME is set, the root group will be parsed from the value of this attribute
+ *     - else the root group will be considered the target group of the application
+ * 2. adds users to the appropriate groups
+ *
+ * @author Dominik Frantisek Bucik <bucik@ics.muni.cz>
+ */
+public class BBMRIResources extends DefaultRegistrarModule {
+
+	private final static Logger log = LoggerFactory.getLogger(BBMRIResources.class);
+
+	// field names
+	private static final String RESOURCE_IDS = "resourceIds";
+	private static final String RESOURCE_ORIGIN = "resourceOrigin";
+
+	// configuration attributes
+	private static final String RESOURCE_ID_ATTR_NAME = "urn:perun:group:attribute-def:def:resourceIDAttrName";
+	private static final String RESOURCE_ORIGIN_ENABLED_ATTR_NAME = "urn:perun:group:attribute-def:def:resourceOriginEnabled";
+	private static final String RESOURCES_ROOT_GROUP_ATTR_NAME = "urn:perun:group:attribute-def:def:resourcesRootGroup";
+
+	/**
+	 * Finds groups representing resources by provided input, adds user into these groups and removes from the
+	 * group where this application form is used.
+	 *
+	 * @param session who approves the application
+	 * @param app application
+	 * @return unchanged application
+	 * @throws PerunException in case of internal error in Perun
+	 */
+	@Override
+	public Application approveApplication(PerunSession session, Application app)
+		throws VoNotExistsException, UserNotExistsException, PrivilegeException,
+		MemberNotExistsException, RegistrarException, GroupNotExistsException,
+		AttributeNotExistsException, WrongAttributeAssignmentException, ExternallyManagedException,
+		WrongAttributeValueException, WrongReferenceAttributeValueException, NotGroupMemberException
+	{
+		if (app.getGroup() == null) {
+			throw new RegistrarException(
+				"Invalid usage of registrar module - module '" + this.getClass().getName() +
+					"' should be used on group level only");
+		}
+
+		// get perun and beans from session
+		PerunBl perun = (PerunBl) session.getPerun();
```

Comment:
I believe you have changed this to the correct type `PerunBl`, but in order to actually call the BL layer you need to also call proper method with `Bl` suffix, eg.: `getMembersManager()` -> `getMembersManagerBl()`.

Reply:
Updated as well. Hopefully this is the last needed change :).